### ISSUE TITLE
Some Python Native int Comparison Fails with np.int64 Max

### ISSUE THREAD

**Issue by mikeronayne** [numpy/numpy#19579]
### Reproducing code example:

```python
import numpy as np
>>> np.iinfo(np.int64).max + 1 == np.int64(np.iinfo(np.int64).max)
True
>>> np.iinfo(np.int64).max + 2 == np.int64(np.iinfo(np.int64).max)
True
>>> np.iinfo(np.int64).max + 1536 == np.int64(np.iinfo(np.int64).max) # 9223372036854777343
True
>>> np.iinfo(np.int64).max + 1537 == np.int64(np.iinfo(np.int64).max)
False
```

After searching `9223372036854777343` on the web, I'm thinking maybe one (or both) of the operands are being casted to float behind the scenes?

### NumPy/Python version information:

1.21.1 3.9.4 (tags/v3.9.4:1f2e308, Apr  6 2021, 13:40:21) [MSC v.1928 64 bit (AMD64)]

**Comment by seberg** [comment]
> After searching 9223372036854777343 on the web, I'm thinking maybe one (or both) of the operands are being casted to float behind the scenes?

Yes, that is unfortunately the case currently.  The reason is that this is the "default" promotion: `np.result_type(np.uint64, np.int64)` gives `float64`.  Arguably, that behaviour is often bad (although maybe often OK for math), and even more so for comparisons.

It could be fixed by adding custom comparisons for these types, although I am not certain what should happen in general.  Maybe we should fix comparisons and then even consider deprecating this promotion entirely (forcing the user to explicitly cast).

See e.g. gh-8851 (and a few issues linked to and from there), which this is a duplicate of.

**Comment by preethamdheeraj** [comment]
Hi, is someone sitll working on this issue. I want to work on this and see if I can develope a fix for solving the issue.
